File: src/Pible/Parser.zig
const std = @import("std");
const Lexer = @import("Lexer.zig");
const Token = Lexer.Token;
pub const TokenType = Lexer.TokenType;

pub const ParseError = error{
    ParseError,
    OutOfMemory,
};

pub const NodeType = enum {
    Program,
    FunctionDecl,
    VarDecl,
    Block,
    ExprStmt,
    ReturnStmt,
    IfStmt,
    WhileStmt,
    BinaryExpr,
    UnaryExpr,
    CallExpr,
    Literal,
    Identifier,
};

pub const Node = struct {
    type: NodeType,
    token: Token,
    children: std.ArrayList(*Node),
    allocator: std.mem.Allocator,

    pub fn init(allocator: std.mem.Allocator, node_type: NodeType, token: Token) ParseError!*Node {
        const node = try allocator.create(Node);
        node.* = .{
            .type = node_type,
            .token = token,
            .children = std.ArrayList(*Node){},
            .allocator = allocator,
        };
        return node;
    }

    pub fn deinit(self: *Node) void {
        for (self.children.items) |child| {
            child.deinit();
        }
        self.children.deinit(self.allocator);
        self.allocator.destroy(self);
    }

    pub fn addChild(self: *Node, child: *Node) ParseError!void {
        try self.children.append(self.allocator, child);
    }
};

pub const Parser = struct {
    tokens: []const Token,
    current: usize = 0,
    allocator: std.mem.Allocator,

    const Self = @This();

    pub fn init(allocator: std.mem.Allocator, tokens: []const Token) Self {
        return .{
            .tokens = tokens,
            .allocator = allocator,
        };
    }

    pub fn parse(self: *Self) ParseError!*Node {
        const program = try Node.init(self.allocator, .Program, Token{
            .type = .Eof,
            .lexeme = "",
            .line = 1,
            .column = 1,
        });
        errdefer program.deinit();

        while (!self.isAtEnd()) {
            if (self.peek().type == .Eof) break;
            const decl = self.declaration() catch |err| {
                // Clean up program node if declaration fails
                return err;
            };
            try program.addChild(decl);
        }

        return program;
    }

    fn declaration(self: *Self) ParseError!*Node {
        // Handle export keyword
        if (self.match(.Export)) {
            return self.declaration(); // Skip export for now, just parse the following declaration
        }
        
        if (self.match(.U0) or self.match(.U8) or self.match(.U16) or 
            self.match(.U32) or self.match(.U64) or self.match(.I8) or 
            self.match(.I16) or self.match(.I32) or self.match(.I64) or self.match(.F64)) {
            
            // Look ahead to determine if this is a function or variable declaration
            if (self.check(.Identifier)) {
                const saved_current = self.current;
                _ = self.advance(); // consume identifier
                
                if (self.check(.LeftParen)) {
                    // It's a function declaration, reset and parse as function
                    self.current = saved_current;
                    return self.functionDeclaration();
                } else {
                    // It's a variable declaration, reset and parse as variable
                    self.current = saved_current;
                    return self.varDeclaration();
                }
            } else {
                return error.ParseError;
            }
        }
        return self.statement();
    }

    /// Parse function parameter list and add parameter nodes to function
    fn parseParameters(self: *Self, func: *Node) ParseError!void {
        if (!self.check(.RightParen)) {
            while (true) {
                // Parameter type must be specified
                if (!self.match(.U0) and !self.match(.U8) and !self.match(.U16) and 
                    !self.match(.U32) and !self.match(.U64) and !self.match(.I8) and 
                    !self.match(.I16) and !self.match(.I32) and !self.match(.I64) and !self.match(.F64)) {
                    return error.ParseError;
                }
                _ = self.previous(); // paramType - consume but not use for now
                
                // Handle pointer types (optional *)
                if (self.match(.Star)) {
                    // This is a pointer type, continue
                }
                
                // Parameter name
                const paramName = try self.consume(.Identifier, "Expected parameter name");
                
                // Create parameter node
                const param = try Node.init(self.allocator, .VarDecl, paramName);
                func.addChild(param) catch |err| {
                    // Clean up parameter node if adding fails
                    param.deinit(self.allocator);
                    return err;
                };
                
                if (!self.match(.Comma)) break;
            }
        }
    }

    fn functionDeclaration(self: *Self) ParseError!*Node {
        _ = self.previous(); // returnType - consume but not use for now
        const name = try self.consume(.Identifier, "Expected function name");
        
        const func = try Node.init(self.allocator, .FunctionDecl, name);
        errdefer func.deinit();
        
        _ = try self.consume(.LeftParen, "Expected '(' after function name");
        
        // Parse parameters using helper method
        try self.parseParameters(func);
        
        _ = try self.consume(.RightParen, "Expected ')' after parameters");
        _ = try self.consume(.LeftBrace, "Expected '{' before function body");
        
        const body = try self.block();
        try func.addChild(body);
        
        return func;
    }

    fn statement(self: *Self) ParseError!*Node {
        if (self.match(.If)) return self.ifStatement();
        if (self.match(.While)) return self.whileStatement();
        if (self.match(.For)) return self.forStatement();
        if (self.match(.Return)) return self.returnStatement();
        if (self.match(.LeftBrace)) return self.block();
        
        // Variable declaration
        if (self.match(.U0) or self.match(.U8) or self.match(.U16) or 
            self.match(.U32) or self.match(.U64) or self.match(.I8) or 
            self.match(.I16) or self.match(.I32) or self.match(.I64) or self.match(.F64)) {
            return self.varDeclaration();
        }
        
        return self.expressionStatement();
    }

    fn ifStatement(self: *Self) ParseError!*Node {
        const ifToken = self.previous();
        const stmt = try Node.init(self.allocator, .IfStmt, ifToken);
        errdefer stmt.deinit();
        
        _ = try self.consume(.LeftParen, "Expected '(' after 'if'");
        const condition = try self.expression();
        try stmt.addChild(condition);
        _ = try self.consume(.RightParen, "Expected ')' after if condition");
        
        const thenBranch = try self.statement();
        try stmt.addChild(thenBranch);
        
        if (self.match(.Else)) {
            const elseBranch = try self.statement();
            try stmt.addChild(elseBranch);
        }
        
        return stmt;
    }

    fn whileStatement(self: *Self) ParseError!*Node {
        const whileToken = self.previous();
        const stmt = try Node.init(self.allocator, .WhileStmt, whileToken);
        errdefer stmt.deinit();
        
        _ = try self.consume(.LeftParen, "Expected '(' after 'while'");
        const condition = try self.expression();
        try stmt.addChild(condition);
        _ = try self.consume(.RightParen, "Expected ')' after while condition");
        
        const body = try self.statement();
        try stmt.addChild(body);
        
        return stmt;
    }

    fn forStatement(self: *Self) ParseError!*Node {
        const forToken = self.previous();
        const stmt = try Node.init(self.allocator, .WhileStmt, forToken); // Use WhileStmt for now
        errdefer stmt.deinit();
        
        _ = try self.consume(.LeftParen, "Expected '(' after 'for'");
        
        // Initializer
        const initializer = if (self.match(.Semicolon)) null
        else if (self.match(.U0) or self.match(.U8) or self.match(.U16) or 
                 self.match(.U32) or self.match(.U64) or self.match(.I8) or 
                 self.match(.I16) or self.match(.I32) or self.match(.I64) or self.match(.F64))
            try self.varDeclaration()
        else try self.expressionStatement();
        
        if (initializer) |init_expr| try stmt.addChild(init_expr);
        
        // Condition
        const condition = if (!self.check(.Semicolon)) try self.expression() else null;
        _ = try self.consume(.Semicolon, "Expected ';' after for loop condition");
        if (condition) |cond| try stmt.addChild(cond);
        
        // Increment
        const increment = if (!self.check(.RightParen)) try self.expression() else null;
        _ = try self.consume(.RightParen, "Expected ')' after for clauses");
        if (increment) |inc| try stmt.addChild(inc);
        
        const body = try self.statement();
        try stmt.addChild(body);
        
        return stmt;
    }

    fn returnStatement(self: *Self) ParseError!*Node {
        const returnToken = self.previous();
        const stmt = try Node.init(self.allocator, .ReturnStmt, returnToken);
        errdefer stmt.deinit();
        
        if (!self.check(.Semicolon)) {
            const value = try self.expression();
            try stmt.addChild(value);
        }
        
        _ = try self.consume(.Semicolon, "Expected ';' after return value");
        return stmt;
    }

    fn varDeclaration(self: *Self) ParseError!*Node {
        _ = self.previous(); // typeToken - consume but not use for now
        const name = try self.consume(.Identifier, "Expected variable name");
        
        const varDecl = try Node.init(self.allocator, .VarDecl, name);
        errdefer varDecl.deinit();
        
        if (self.match(.Equal)) {
            const initializer = try self.expression();
            try varDecl.addChild(initializer);
        }
        
        _ = try self.consume(.Semicolon, "Expected ';' after variable declaration");
        return varDecl;
    }

    fn block(self: *Self) ParseError!*Node {
        const blockNode = try Node.init(self.allocator, .Block, Token{
            .type = .LeftBrace,
            .lexeme = "{",
            .line = self.peek().line,
            .column = self.peek().column,
        });
        errdefer blockNode.deinit();
        
        while (!self.check(.RightBrace) and !self.isAtEnd()) {
            const stmt = try self.declaration();
            try blockNode.addChild(stmt);
        }
        
        _ = try self.consume(.RightBrace, "Expected '}' after block");
        return blockNode;
    }

    fn expressionStatement(self: *Self) ParseError!*Node {
        const expr = try self.expression();
        _ = try self.consume(.Semicolon, "Expected ';' after expression");
        
        const exprStmt = try Node.init(self.allocator, .ExprStmt, Token{
            .type = .Semicolon,
            .lexeme = ";",
            .line = self.previous().line,
            .column = self.previous().column,
        });
        errdefer exprStmt.deinit();
        try exprStmt.addChild(expr);
        return exprStmt;
    }

    fn expression(self: *Self) ParseError!*Node {
        return self.logical_or();
    }

    fn logical_or(self: *Self) ParseError!*Node {
        var expr = try self.logical_and();
        
        while (self.match(.Or)) {
            const operator = self.previous();
            const right = try self.logical_and();
            const binary = try Node.init(self.allocator, .BinaryExpr, operator);
            errdefer binary.deinit();
            try binary.addChild(expr);
            try binary.addChild(right);
            expr = binary;
        }
        
        return expr;
    }

    fn logical_and(self: *Self) ParseError!*Node {
        var expr = try self.equality();
        
        while (self.match(.And)) {
            const operator = self.previous();
            const right = try self.equality();
            const binary = try Node.init(self.allocator, .BinaryExpr, operator);
            errdefer binary.deinit();
            try binary.addChild(expr);
            try binary.addChild(right);
            expr = binary;
        }
        
        return expr;
    }

    fn equality(self: *Self) ParseError!*Node {
        var expr = try self.comparison();
        
        while (self.match(.BangEqual) or self.match(.EqualEqual)) {
            const operator = self.previous();
            const right = try self.comparison();
            const binary = try Node.init(self.allocator, .BinaryExpr, operator);
            errdefer binary.deinit();
            try binary.addChild(expr);
            try binary.addChild(right);
            expr = binary;
        }
        
        return expr;
    }

    fn comparison(self: *Self) ParseError!*Node {
        var expr = try self.term();
        
        while (self.match(.Greater) or self.match(.GreaterEqual) or 
               self.match(.Less) or self.match(.LessEqual)) {
            const operator = self.previous();
            const right = try self.term();
            const binary = try Node.init(self.allocator, .BinaryExpr, operator);
            errdefer binary.deinit();
            try binary.addChild(expr);
            try binary.addChild(right);
            expr = binary;
        }
        
        return expr;
    }

    fn term(self: *Self) ParseError!*Node {
        var expr = try self.factor();
        
        while (self.match(.Minus) or self.match(.Plus)) {
            const operator = self.previous();
            const right = try self.factor();
            const binary = try Node.init(self.allocator, .BinaryExpr, operator);
            errdefer binary.deinit();
            try binary.addChild(expr);
            try binary.addChild(right);
            expr = binary;
        }
        
        return expr;
    }

    fn factor(self: *Self) ParseError!*Node {
        var expr = try self.unary();
        
        while (self.match(.Slash) or self.match(.Star)) {
            const operator = self.previous();
            const right = try self.unary();
            const binary = try Node.init(self.allocator, .BinaryExpr, operator);
            errdefer binary.deinit();
            try binary.addChild(expr);
            try binary.addChild(right);
            expr = binary;
        }
        
        return expr;
    }

    fn unary(self: *Self) ParseError!*Node {
        if (self.match(.Bang) or self.match(.Minus)) {
            const operator = self.previous();
            const right = try self.unary();
            const unary_expr = try Node.init(self.allocator, .UnaryExpr, operator);
            errdefer unary_expr.deinit();
            try unary_expr.addChild(right);
            return unary_expr;
        }
        
        return self.call();
    }

    fn call(self: *Self) ParseError!*Node {
        var expr = try self.primary();
        
        while (true) {
            if (self.match(.LeftParen)) {
                expr = try self.finishCall(expr);
            } else {
                break;
            }
        }
        
        return expr;
    }

    fn finishCall(self: *Self, callee: *Node) ParseError!*Node {
        const call_expr = try Node.init(self.allocator, .CallExpr, Token{
            .type = .LeftParen,
            .lexeme = "(",
            .line = self.previous().line,
            .column = self.previous().column,
        });
        errdefer call_expr.deinit();
        try call_expr.addChild(callee);
        
        if (!self.check(.RightParen)) {
            while (true) {
                const arg = try self.expression();
                try call_expr.addChild(arg);
                if (!self.match(.Comma)) break;
            }
        }
        
        _ = try self.consume(.RightParen, "Expected ')' after arguments");
        return call_expr;
    }

    fn primary(self: *Self) ParseError!*Node {
        if (self.match(.True) or self.match(.False)) {
            return Node.init(self.allocator, .Literal, self.previous());
        }
        
        if (self.match(.NumberLiteral)) {
            return Node.init(self.allocator, .Literal, self.previous());
        }
        
        if (self.match(.StringLiteral)) {
            return Node.init(self.allocator, .Literal, self.previous());
        }
        
        if (self.match(.Identifier) or self.match(.PrintF)) {
            return Node.init(self.allocator, .Identifier, self.previous());
        }
        
        if (self.match(.LeftParen)) {
            const expr = try self.expression();
            _ = try self.consume(.RightParen, "Expected ')' after expression");
            return expr;
        }
        
        return error.ParseError;
    }

    // Helper methods
    fn match(self: *Self, tokenType: TokenType) bool {
        if (self.check(tokenType)) {
            _ = self.advance();
            return true;
        }
        return false;
    }

    fn check(self: *Self, tokenType: TokenType) bool {
        if (self.isAtEnd()) return false;
        return self.peek().type == tokenType;
    }

    fn advance(self: *Self) Token {
        if (!self.isAtEnd()) self.current += 1;
        return self.previous();
    }

    fn isAtEnd(self: *Self) bool {
        return self.peek().type == .Eof;
    }

    fn peek(self: *Self) Token {
        return self.tokens[self.current];
    }

    fn previous(self: *Self) Token {
        return self.tokens[self.current - 1];
    }

    fn consume(self: *Self, tokenType: TokenType, message: []const u8) ParseError!Token {
        if (self.check(tokenType)) return self.advance();
        
        std.debug.print("Parse error at line {d}: {s}\n", .{ self.peek().line, message });
        return error.ParseError;
    }
};

File: src/pible/parser.rs
use thiserror::Error;
use crate::pible::lexer::{Token, TokenType};

#[derive(Error, Debug)]
pub enum ParseError {
    #[error("Unexpected token: {0:?} at line {1}")]
    UnexpectedToken(TokenType, usize),
    #[error("Expected token: {expected:?}, found: {found:?} at line {line}")]
    ExpectedToken { expected: TokenType, found: TokenType, line: usize },
    #[error("Unexpected end of file")]
    UnexpectedEof,
}

#[derive(Debug, Clone, PartialEq)]
pub enum NodeType {
    Program,
    FunctionDecl,
    Block,
    Statement,
    Expression,
    Identifier,
    Literal,
}

#[derive(Debug, Clone)]
pub struct Node {
    pub node_type: NodeType,
    pub value: Option<String>,
    pub children: Vec<Node>,
}

impl Node {
    pub fn new(node_type: NodeType) -> Self {
        Self {
            node_type,
            value: None,
            children: Vec::new(),
        }
    }

    pub fn with_value(node_type: NodeType, value: String) -> Self {
        Self {
            node_type,
            value: Some(value),
            children: Vec::new(),
        }
    }

    pub fn add_child(&mut self, child: Node) {
        self.children.push(child);
    }
}

pub struct Parser<'a> {
    tokens: Vec<Token<'a>>,
    current: usize,
}

impl<'a> Parser<'a> {
    pub fn new(tokens: Vec<Token<'a>>) -> Self {
        Self {
            tokens,
            current: 0,
        }
    }

    pub fn parse(&mut self) -> Result<Node, ParseError> {
        let mut program = Node::new(NodeType::Program);
        
        while !self.is_at_end() {
            if let Ok(declaration) = self.declaration() {
                program.add_child(declaration);
            } else {
                // Skip to next declaration on error
                self.synchronize();
            }
        }

        Ok(program)
    }

    fn declaration(&mut self) -> Result<Node, ParseError> {
        if self.match_token(&[TokenType::Export]) {
            self.function_declaration()
        } else if self.check(&TokenType::U0) || self.check(&TokenType::U8) || 
                  self.check(&TokenType::U16) || self.check(&TokenType::U32) || 
                  self.check(&TokenType::U64) {
            self.function_declaration()
        } else {
            self.statement()
        }
    }

    fn function_declaration(&mut self) -> Result<Node, ParseError> {
        // Parse return type
        let return_type_str = self.advance().lexeme.to_string();
        
        // Parse function name
        let name_token = self.consume(TokenType::Identifier, "Expected function name")?;
        let name_str = name_token.lexeme.to_string();
        
        // Parse parameters
        self.consume(TokenType::LeftParen, "Expected '(' after function name")?;
        let mut params = Vec::new();
        
        if !self.check(&TokenType::RightParen) {
            loop {
                // Parse parameter type and name
                let param_type_str = self.advance().lexeme.to_string();
                if self.previous().token_type == TokenType::Star {
                    // Handle pointer types like U8*
                    let _ = self.advance(); // consume the base type
                }
                if self.check(&TokenType::Identifier) {
                    let param_name_token = self.advance();
                    let mut param_node = Node::with_value(NodeType::Identifier, param_name_token.lexeme.to_string());
                    param_node.value = Some(param_type_str);
                    params.push(param_node);
                }
                
                if !self.match_token(&[TokenType::Comma]) {
                    break;
                }
            }
        }
        self.consume(TokenType::RightParen, "Expected ')' after parameters")?;
        
        // Parse function body
        let body = self.block_statement()?;
        
        let mut function = Node::with_value(NodeType::FunctionDecl, format!("{}:{}", return_type_str, name_str));
        for param in params {
            function.add_child(param);
        }
        function.add_child(body);
        
        Ok(function)
    }

    fn statement(&mut self) -> Result<Node, ParseError> {
        if self.match_token(&[TokenType::Return]) {
            self.return_statement()
        } else if self.match_token(&[TokenType::LeftBrace]) {
            self.block_statement()
        } else {
            self.expression_statement()
        }
    }

    fn return_statement(&mut self) -> Result<Node, ParseError> {
        let mut stmt = Node::with_value(NodeType::Statement, "return".to_string());
        
        if !self.check(&TokenType::Semicolon) {
            let expr = self.expression()?;
            stmt.add_child(expr);
        }
        
        self.consume(TokenType::Semicolon, "Expected ';' after return value")?;
        Ok(stmt)
    }

    fn block_statement(&mut self) -> Result<Node, ParseError> {
        let mut block = Node::new(NodeType::Block);
        
        while !self.check(&TokenType::RightBrace) && !self.is_at_end() {
            let stmt = self.declaration()?;
            block.add_child(stmt);
        }
        
        self.consume(TokenType::RightBrace, "Expected '}' after block")?;
        Ok(block)
    }

    fn expression_statement(&mut self) -> Result<Node, ParseError> {
        let expr = self.expression()?;
        self.consume(TokenType::Semicolon, "Expected ';' after expression")?;
        Ok(expr)
    }

    fn expression(&mut self) -> Result<Node, ParseError> {
        self.call()
    }

    fn call(&mut self) -> Result<Node, ParseError> {
        let mut expr = self.primary()?;
        
        while self.match_token(&[TokenType::LeftParen]) {
            expr = self.finish_call(expr)?;
        }
        
        Ok(expr)
    }

    fn finish_call(&mut self, callee: Node) -> Result<Node, ParseError> {
        let mut call = Node::with_value(NodeType::Expression, "call".to_string());
        call.add_child(callee);
        
        if !self.check(&TokenType::RightParen) {
            loop {
                let arg = self.expression()?;
                call.add_child(arg);
                
                if !self.match_token(&[TokenType::Comma]) {
                    break;
                }
            }
        }
        
        self.consume(TokenType::RightParen, "Expected ')' after arguments")?;
        Ok(call)
    }

    fn primary(&mut self) -> Result<Node, ParseError> {
        if self.match_token(&[TokenType::True, TokenType::False]) {
            let token_str = self.previous().lexeme.to_string();
            let mut node = Node::new(NodeType::Literal);
            node.value = Some(token_str);
            Ok(node)
        } else if self.match_token(&[TokenType::NumberLiteral]) {
            let token_str = self.previous().lexeme.to_string();
            let mut node = Node::new(NodeType::Literal);
            node.value = Some(token_str);
            Ok(node)
        } else if self.match_token(&[TokenType::StringLiteral]) {
            let token_str = self.previous().lexeme.to_string();
            let mut node = Node::new(NodeType::Literal);
            node.value = Some(token_str);
            Ok(node)
        } else if self.match_token(&[TokenType::Identifier, TokenType::PrintF]) {
            let token_str = self.previous().lexeme.to_string();
            let mut node = Node::new(NodeType::Identifier);
            node.value = Some(token_str);
            Ok(node)
        } else if self.match_token(&[TokenType::LeftParen]) {
            let expr = self.expression()?;
            self.consume(TokenType::RightParen, "Expected ')' after expression")?;
            Ok(expr)
        } else {
            Err(ParseError::UnexpectedToken(
                self.peek().token_type,
                self.peek().line
            ))
        }
    }

    fn match_token(&mut self, types: &[TokenType]) -> bool {
        for token_type in types {
            if self.check(token_type) {
                self.advance();
                return true;
            }
        }
        false
    }

    fn check(&self, token_type: &TokenType) -> bool {
        if self.is_at_end() {
            false
        } else {
            &self.peek().token_type == token_type
        }
    }

    fn advance(&mut self) -> &Token<'a> {
        if !self.is_at_end() {
            self.current += 1;
        }
        self.previous()
    }

    fn is_at_end(&self) -> bool {
        self.peek().token_type == TokenType::Eof
    }

    fn peek(&self) -> &Token<'a> {
        &self.tokens[self.current]
    }

    fn previous(&self) -> &Token<'a> {
        &self.tokens[self.current - 1]
    }

    fn consume(&mut self, token_type: TokenType, _message: &str) -> Result<&Token<'a>, ParseError> {
        if self.check(&token_type) {
            Ok(self.advance())
        } else {
            Err(ParseError::ExpectedToken {
                expected: token_type,
                found: self.peek().token_type,
                line: self.peek().line,
            })
        }
    }

    fn synchronize(&mut self) {
        self.advance();

        while !self.is_at_end() {
            if self.previous().token_type == TokenType::Semicolon {
                return;
            }

            match self.peek().token_type {
                TokenType::Class | TokenType::For | TokenType::If | 
                TokenType::While | TokenType::Return => return,
                _ => {}
            }

            self.advance();
        }
    }
}